### PR TITLE
Lazily gather the lists of available sls data from the master

### PR DESCRIPTION
**Pull request by sathieu**
### What does this PR do?

This improves significantly the performance of a state.apply when using
GitFS and a lot of branches.

### What issues does this PR fix or reference?
none

### Previous Behavior
3 minutes with 276 branches.

### New Behavior
25 seconds

### Tests written?

No. Already covered by existing tests.

### Commits signed with GPG?

No

### CODE REVIEW

**Comment by sathieu** [comment]
Note : I have tested a similar patch for pillars (in `salt/pillar/__init__.py:462` `Pillar.__gather_avail()`), but the performance improvement is not that significant. Probably because the loop in master-only, while the loop in `BaseHighState` is on the minion but talks to the master for each env.

**Comment by sathieu** [comment]
@twangboy Could you please review?

**Comment by sathieu** [comment]
@dwoz I've added tests

File: salt/state.py [hunk 1]
```diff
@@ -362,7 +362,6 @@ class StateError(Exception):
     '''
     Custom exception class.
     '''
-    pass
```
Comment (twangboy) [inline salt/state.py]: Isn't this supposed to remain?

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/state.py (hunk 1 above)
Comment (sathieu) [reply salt/state.py]: As written in 332de26df886a442ccbfa43ccce1d73dd1b3bf96, this is to fix:
```
salt/state.py:370: [W0107(unnecessary-pass), StateError] Unnecessary pass statement
```

Ref:
- http://pylint.pycqa.org/en/latest/technical_reference/features.html
- https://help.semmle.com/wiki/display/PYTHON/Unnecessary+pass

**Review by DmitryKuzmenko — DISMISSED** [review]

**Comment by waynew** [comment]
@thatch45 can you take a look at this? It *looks* pretty straightforward to me but I'm not 100% familiar with the implications of futzing with the states.

**Review by thatch45 — DISMISSED** [review]

**Comment by thatch45** [comment]
I really like this approach, it is clean, and it can be pulled out if there are issues. But this does look like it should work.
I love simple, clean additions :)

**Comment by DmitryKuzmenko** [comment]
@twangboy your turn!

**Comment by DmitryKuzmenko** [comment]
@sathieu could you please resolve merge conflicts?

**Comment by DmitryKuzmenko** [comment]
@sathieu could you please resolve the merge conflicts and run the pre-commit script on your code to re-format it with `isort` and `black` we're using now.
```
pip install pre-commit
pre-commit install
FILES="$(git show --pretty="" --name-only master..HEAD)"
pre-commit run isort $FILES
pre-commit run black $FILES
```
next time you'll commit it will be ran automatically.

**Comment by sathieu** [comment]
@DmitryKuzmenko I've rebased and used pre-commit hook. I had to use `virtualenv -p /usr/bin/python3 DESTDIR`, otherwise python2 was used.

Then, it failed with:
```
  Attempting uninstall: wrapt
    Found existing installation: wrapt 1.10.11
ERROR: Cannot uninstall 'wrapt'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.
Session lint-tests-pre-commit failed.
```
I've commited again (black and isort already runt) with `--no-verify` and pushed here.

**Comment by DmitryKuzmenko** [comment]
@sathieu thank you for report. We'll work on it.
Meanwhile I'll check your code for you. Thank you for contribution.

**Comment by waynew** [comment]
@sathieu did you use `pip` or `python3 -m pip install --user pre-commit`?  It *should* be [defaulting to python3](https://github.com/saltstack/salt/pull/56586/files#diff-82c3b36709926de3a1b27bea0d8258ccR2) when installing the hooks.

If just installing pre-commit that way doesn't work running `rm -rf ~/.cache/pre-commit/` may help.

**Comment by DmitryKuzmenko** [comment]
@waynew ah, it looks it's my fault. I should recommend `python3 pip` next time.

**Review by dwoz — APPROVED** [review]

**Pull request merged**